rust/src/trans/sum/mod.rs

```diff
@@ -1,62 +1,106 @@
-#[cfg(feature="ffi")]
+#[cfg(feature = "ffi")]
 mod ffi;
 
-use std::iter::Sum;
-use std::ops::Sub;
+mod int;
+pub use int::*;
 
-use num::Zero;
+mod float;
+pub use float::*;
 
-use crate::core::{Function, StabilityRelation, Transformation};
-use crate::dist::{AbsoluteDistance, IntDistance, SymmetricDistance};
+use crate::core::Transformation;
+use crate::dist::{AbsoluteDistance, SymmetricDistance};
 use crate::dom::{AllDomain, BoundedDomain, SizedDomain, VectorDomain};
 use crate::error::*;
-use crate::traits::{CheckNull, DistanceConstant, ExactIntCast, InfCast, InfDiv, SaturatingAdd, InfSub, AlertingAbs};
-
-pub fn make_bounded_sum<T>(
-    bounds: (T, T)
-) -> Fallible<Transformation<VectorDomain<BoundedDomain<T>>, AllDomain<T>, SymmetricDistance, AbsoluteDistance<T>>>
-    where T: DistanceConstant<IntDistance> + Sub<Output=T> + SaturatingAdd + Zero + CheckNull + AlertingAbs,
-          IntDistance: InfCast<T> {
-    let (lower, upper) = bounds.clone();
-
-    Ok(Transformation::new(
-        VectorDomain::new(BoundedDomain::new_closed(bounds)?),
-        AllDomain::new(),
-        Function::new(|arg: &Vec<T>| arg.iter().fold(T::zero(), |sum, v| sum.saturating_add(v))),
-        SymmetricDistance::default(),
-        AbsoluteDistance::default(),
-        StabilityRelation::new_from_constant(
-            lower.alerting_abs()?.total_max(upper.alerting_abs()?)?)))
+use crate::traits::{CheckNull, TotalOrd};
+use crate::trans::make_ordered_random;
+
+pub fn make_bounded_sum<T: MakeBoundedSum>(bounds: (T, T)) -> Fallible<BoundedSumTrans<T>> {
+    T::make_bounded_sum(bounds)
+}
+pub fn make_sized_bounded_sum<T: MakeSizedBoundedSum>(
+    size: usize,
+    bounds: (T, T),
+) -> Fallible<SizedBoundedSumTrans<T>> {
+    T::make_sized_bounded_sum(size, bounds)
+}
+
+// implementations delegate to:
+// make_(sized_)?bounded_sum
+// make_(sized_)?bounded_int_(checked|monotonic|ordered|split)_sum
+// make_(sized_)?bounded_float_(checked|ordered)_sum
+
+type BoundedSumTrans<T> = Transformation<
+    VectorDomain<BoundedDomain<T>>,
+    AllDomain<T>,
+    SymmetricDistance,
+    AbsoluteDistance<T>,
+>;
+
+pub trait MakeBoundedSum: Sized + CheckNull + Clone + TotalOrd {
+    fn make_bounded_sum(bounds: (Self, Self)) -> Fallible<BoundedSumTrans<Self>>;
+}
+
+macro_rules! impl_make_bounded_sum_int {
+    ($($ty:ty)+) => ($(impl MakeBoundedSum for $ty {
+        fn make_bounded_sum(bounds: (Self, Self)) -> Fallible<BoundedSumTrans<Self>> {
+            make_bounded_int_monotonic_sum(bounds)
+                .or_else(|_| make_bounded_int_split_sum(bounds))
```

Comment:
The additional dispatch options here (and in `impl_make_sized_bounded_sum_int`) are nice. But basing things on the constructors returning `Err` feels a little risky, as unexpected failures would be masked. Would it be possible to have explicit choice logic here, rather than relying on errors on the happy path?

Reply:
I've updated to have explicit logic.